`make check` fails to build with GCC 10

I am trying to build and run SWIG tests at Arch Linux. Arch has GCC10 and it looks like it introduces some compilation issues:

```
Note: 'swig -<lang> -help' displays options for a specific target language.

Supported Target Language Options
     -csharp         - Generate C# wrappers
     -d              - Generate D wrappers
     -go             - Generate Go wrappers
     -guile          - Generate Guile wrappers
     -java           - Generate Java wrappers
     -javascript     - Generate Javascript wrappers
     -lua            - Generate Lua wrappers
     -octave         - Generate Octave wrappers
     -perl5          - Generate Perl 5 wrappers
     -php7           - Generate PHP 7 wrappers
     -python         - Generate Python wrappers
     -r              - Generate R (aka GNU S) wrappers
     -ruby           - Generate Ruby wrappers
     -scilab         - Generate Scilab wrappers
     -tcl8           - Generate Tcl 8 wrappers
     -xml            - Generate XML wrappers

Experimental Target Language Options
     -mzscheme       - Generate MzScheme/Racket wrappers
     -ocaml          - Generate OCaml wrappers

General Options
     -addextern      - Add extra extern declarations
     -c++            - Enable C++ processing
     -co <file>      - Check <file> out of the SWIG library
     -copyctor       - Automatically generate copy constructors wherever possible
     -cpperraswarn   - Treat the preprocessor #error statement as #warning (default)
     -cppext <ext>   - Change file extension of generated C++ files to <ext>
                       (default is cxx)
     -copyright      - Display copyright notices
     -debug-classes  - Display information about the classes found in the interface
     -debug-module <n>- Display module parse tree at stages 1-4, <n> is a csv list of stages
     -debug-symtabs  - Display symbol tables information
     -debug-symbols  - Display target language symbols in the symbol tables
     -debug-csymbols - Display C symbols in the symbol tables
     -debug-lsymbols - Display target language layer symbols
     -debug-tags     - Display information about the tags found in the interface
     -debug-template - Display information for debugging templates
     -debug-top <n>  - Display entire parse tree at stages 1-4, <n> is a csv list of stages
     -debug-typedef  - Display information about the types and typedefs in the interface
     -debug-typemap  - Display typemap debugging information
     -debug-tmsearch - Display typemap search debugging information
     -debug-tmused   - Display typemaps used debugging information
     -directors      - Turn on director mode for all the classes, mainly for testing
     -dirprot        - Turn on wrapping of protected members for director classes (default)
     -D<symbol>      - Define a symbol <symbol> (for conditional compilation)
     -E              - Preprocess only, does not generate wrapper code
     -external-runtime [file] - Export the SWIG runtime stack
     -fakeversion <v>- Make SWIG fake the program version number to <v>
     -fcompact       - Compile in compact mode
     -features <list>- Set global features, where <list> is a comma separated list of
                       features, eg -features directors,autodoc=1
                       If no explicit value is given to the feature, a default of 1 is used
     -fastdispatch   - Enable fast dispatch mode to produce faster overload dispatcher code
     -Fmicrosoft     - Display error/warning messages in Microsoft format
     -Fstandard      - Display error/warning messages in commonly used format
     -fvirtual       - Compile in virtual elimination mode
     -help           - Display help
     -I-             - Don't search the current directory
     -I<dir>         - Look for SWIG files in directory <dir>
     -ignoremissing  - Ignore missing include files
     -importall      - Follow all #include statements as imports
     -includeall     - Follow all #include statements
     -l<ifile>       - Include SWIG library file <ifile>
     -macroerrors    - Report errors inside macros
     -makedefault    - Create default constructors/destructors (the default)
     -M              - List all dependencies
     -MD             - Is equivalent to `-M -MF <file>', except `-E' is not implied
     -MF <file>      - Generate dependencies into <file> and continue generating wrappers
     -MM             - List dependencies, but omit files in SWIG library
     -MMD            - Like `-MD', but omit files in SWIG library
     -module <name>  - Set module name to <name>
     -MP             - Generate phony targets for all dependencies
     -MT <target>    - Set the target of the rule emitted by dependency generation
     -nocontract     - Turn off contract checking
     -nocpperraswarn - Do not treat the preprocessor #error statement as #warning
     -nodefault      - Do not generate default constructors nor default destructors
     -nodefaultctor  - Do not generate implicit default constructors
     -nodefaultdtor  - Do not generate implicit default destructors
     -nodirprot      - Do not wrap director protected members
     -noexcept       - Do not wrap exception specifiers
     -nofastdispatch - Disable fast dispatch mode (default)
     -nopreprocess   - Skip the preprocessor step
     -notemplatereduce - Disable reduction of the typedefs in templates
     -O              - Enable the optimization options:
                        -fastdispatch -fvirtual
     -o <outfile>    - Set name of C/C++ output file to <outfile>
     -oh <headfile>  - Set name of C++ output header file for directors to <headfile>
     -outcurrentdir  - Set default output dir to current dir instead of input file's path
     -outdir <dir>   - Set language specific files output directory to <dir>
     -pcreversion    - Display PCRE version information
     -small          - Compile in virtual elimination and compact mode
     -swiglib        - Report location of SWIG library and exit
     -templatereduce - Reduce all the typedefs in templates
     -v              - Run in verbose mode
     -version        - Display SWIG version number
     -Wall           - Remove all warning suppression, also implies -Wextra
     -Wallkw         - Enable keyword warnings for all the supported languages
     -Werror         - Treat warnings as errors
     -Wextra         - Adds the following additional warnings: 202,309,403,405,512,321,322
     -w<list>        - Suppress/add warning messages, eg -w401,+321 - see Warnings.html
     -xmlout <file>  - Write XML version of the parse tree to <file> after normal processing

Options can also be defined using the SWIG_FEATURES environment variable, for example:

  $ SWIG_FEATURES="-Wall"
  $ export SWIG_FEATURES
  $ swig -python interface.i

is equivalent to:

  $ swig -Wall -python interface.i

Arguments may also be passed in a file, separated by whitespace. For example:

  $ echo "-Wall -python interface.i" > args.txt
  $ swig @args.txt

Javascript Options (available with -javascript)
     -jsc                   - creates a JavascriptCore extension 
     -v8                    - creates a v8 extension 
     -node                  - creates a node.js extension 
     -debug-codetemplates   - generates information about the origin of code templates

Note: 'swig -<lang> -help' displays options for a specific target language.

test -z "1" || (cd CCache && make check)
make[1]: Entering directory '/mnt/cold/sources/swig/CCache'
SWIG_LIB='.././../Lib' PATH=../..:$PATH SWIG='swig' CC='gcc' NOSOFTLINKSTEST='' CCACHE='../ccache-swig' CCACHE_PROG=ccache-swig ./test.sh
starting testsuite base
starting swig testsuite base
starting testsuite link
starting swig testsuite link
starting testsuite hardlink
starting swig testsuite hardlink
starting testsuite cpp2
starting swig testsuite cpp2
starting testsuite nlevels4
starting testsuite nlevels1
test done - OK
make[1]: Leaving directory '/mnt/cold/sources/swig/CCache'
showing tcl version
8.6
showing perl5 version
This is perl 5, version 30, subversion 2 (v5.30.2) built for x86_64-linux-thread-multi
showing python version
Python 3.8.3
showing java version
openjdk version "1.8.0_252"
OpenJDK Runtime Environment (build 1.8.0_252-b09)
OpenJDK 64-Bit Server VM (build 25.252-b09, mixed mode)
javac 1.8.0_252
showing javascript version
Node.js: v14.3.0
node-gyp: v6.1.0
skipping android version
showing guile version
guile (GNU Guile) 2.2.6
skipping mzscheme version
showing ruby version
ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-linux]
skipping ocaml version
skipping octave version
showing php version
PHP 7.4.6 (cli) (built: May 12 2020 15:48:23) ( NTS )
skipping csharp version
showing lua version
Lua 5.3.5  Copyright (C) 1994-2018 Lua.org, PUC-Rio
skipping r version
skipping scilab version
showing go version
go version go1.14.3 linux/amd64
skipping d version
checking Examples/tcl/class
In file included from /usr/include/c++/10.1.0/ext/string_conversions.h:41,
                 from /usr/include/c++/10.1.0/bits/basic_string.h:6535,
                 from /usr/include/c++/10.1.0/string:55,
                 from /usr/include/c++/10.1.0/stdexcept:39,
                 from example_wrap.cxx:1675:
/usr/include/c++/10.1.0/cstdlib:75:15: fatal error: stdlib.h: No such file or directory
   75 | #include_next <stdlib.h>
      |               ^~~~~~~~~~
compilation terminated.
make[3]: *** [../../Makefile:197: tcl_cpp] Error 1
make[2]: *** [Makefile:13: build] Error 2
make[2]: Target 'check' not remade because of errors.
make[1]: *** [Makefile:216: class.actionexample] Error 2
checking Examples/tcl/constants
checking Examples/tcl/contract
checking Examples/tcl/enum
In file included from /usr/include/c++/10.1.0/ext/string_conversions.h:41,
                 from /usr/include/c++/10.1.0/bits/basic_string.h:6535,
                 from /usr/include/c++/10.1.0/string:55,
                 from /usr/include/c++/10.1.0/stdexcept:39,
                 from example_wrap.cxx:1673:
/usr/include/c++/10.1.0/cstdlib:75:15: fatal error: stdlib.h: No such file or directory
   75 | #include_next <stdlib.h>
      |               ^~~~~~~~~~
compilation terminated.
make[3]: *** [../../Makefile:197: tcl_cpp] Error 1
make[2]: *** [Makefile:13: build] Error 2
make[2]: Target 'check' not remade because of errors.
make[1]: *** [Makefile:216: enum.actionexample] Error 2
checking Examples/tcl/funcptr
checking Examples/tcl/import
In file included from /usr/include/c++/10.1.0/ext/string_conversions.h:41,
                 from /usr/include/c++/10.1.0/bits/basic_string.h:6535,
                 from /usr/include/c++/10.1.0/string:55,
                 from /usr/include/c++/10.1.0/stdexcept:39,
                 from base_wrap.cxx:1673:
/usr/include/c++/10.1.0/cstdlib:75:15: fatal error: stdlib.h: No such file or directory
   75 | #include_next <stdlib.h>
      |               ^~~~~~~~~~
compilation terminated.
make[3]: *** [../../Makefile:197: tcl_cpp] Error 1
make[2]: *** [Makefile:11: build] Error 2
make[2]: Target 'check' not remade because of errors.
make[1]: *** [Makefile:216: import.actionexample] Error 2
checking Examples/tcl/multimap
checking Examples/tcl/operator
In file included from /usr/include/c++/10.1.0/ext/string_conversions.h:41,
                 from /usr/include/c++/10.1.0/bits/basic_string.h:6535,
                 from /usr/include/c++/10.1.0/string:55,
                 from /usr/include/c++/10.1.0/stdexcept:39,
                 from example_wrap.cxx:1673:
/usr/include/c++/10.1.0/cstdlib:75:15: fatal error: stdlib.h: No such file or directory
   75 | #include_next <stdlib.h>
      |               ^~~~~~~~~~
compilation terminated.
make[3]: *** [../../Makefile:197: tcl_cpp] Error 1
make[2]: *** [Makefile:13: build] Error 2
make[2]: Target 'check' not remade because of errors.
make[1]: *** [Makefile:216: operator.actionexample] Error 2
checking Examples/tcl/pointer
checking Examples/tcl/reference
In file included from /usr/include/c++/10.1.0/ext/string_conversions.h:41,
                 from /usr/include/c++/10.1.0/bits/basic_string.h:6535,
                 from /usr/include/c++/10.1.0/string:55,
                 from /usr/include/c++/10.1.0/stdexcept:39,
                 from example_wrap.cxx:1674:
/usr/include/c++/10.1.0/cstdlib:75:15: fatal error: stdlib.h: No such file or directory
   75 | #include_next <stdlib.h>
      |               ^~~~~~~~~~
compilation terminated.
make[3]: *** [../../Makefile:197: tcl_cpp] Error 1
make[2]: *** [Makefile:13: build] Error 2
make[2]: Target 'check' not remade because of errors.
make[1]: *** [Makefile:216: reference.actionexample] Error 2
checking Examples/tcl/simple
checking Examples/tcl/std_vector
In file included from /usr/include/c++/10.1.0/ext/string_conversions.h:41,
                 from /usr/include/c++/10.1.0/bits/basic_string.h:6535,
                 from /usr/include/c++/10.1.0/string:55,
                 from /usr/include/c++/10.1.0/stdexcept:39,
                 from example_wrap.cxx:1683:
/usr/include/c++/10.1.0/cstdlib:75:15: fatal error: stdlib.h: No such file or directory
   75 | #include_next <stdlib.h>
      |               ^~~~~~~~~~
compilation terminated.
make[3]: *** [../../Makefile:197: tcl_cpp] Error 1
make[2]: *** [Makefile:14: build] Error 2
make[2]: Target 'check' not remade because of errors.
make[1]: *** [Makefile:216: std_vector.actionexample] Error 2
checking Examples/tcl/value
checking Examples/tcl/variables
make: *** [Makefile:201: check-tcl-examples] Error 2
```

Hi @wsfulton thank you for your quick reply. The GCC10 is fine. I use it extensively and compiled multiple large projects recently. So most likely the problem is somewhere else.

I briefly googled for this error message and among other things I found this GCC bug https://gcc.gnu.org/bugzilla/show_bug.cgi?id=70129 where @jakubjelinek said:

```
So, IMHO just the packages that use this should either know what they are doing and put the C++ STL system headers first, or don't use STL, or don't mess with -isystem for the default directories.  The last one preferred.
```

So for sake of an experiment I made following quick check

```patch
--- a/configure.ac
+++ b/configure.ac
@@ -118,7 +118,7 @@ AC_SUBST(BOOST_CPPFLAGS)
 dnl How to specify include directories that may be system directories.
 # -I should not be used on system directories (GCC)
 if test "$GCC" = yes; then
-    ISYSTEM="-isystem "
+    ISYSTEM="-I"
 else
     ISYSTEM="-I"
 fi
```

And this particular error message went away. But I do not know SWIG build system well enough to judge whether it is a correct fix or not.

And even with the change above I still get
```
checking Examples/python/varargs
example_wrap.c:154:11: fatal error: Python.h: No such file or directory
  154 | # include <Python.h>
      |           ^~~~~~~~~~
```
So probably more changes related to `-I` handling is needed.

@vadz thank you for the help. If I remove `-s` from `Makegile.in` and re-run `make check` I get following log:

```
make[2]: Entering directory '/home/anatol/sources/swig/Examples/tcl/std_vector'
make -f ../../Makefile SRCDIR='' CXXSRCS='' \
SWIG_LIB_DIR='../../../Lib' SWIGEXE='../../../swig' \
TARGET='example' INTERFACE='example.i' tcl_cpp
make[3]: Entering directory '/home/anatol/sources/swig/Examples/tcl/std_vector'
env SWIG_LIB=../../../Lib  ../../../swig -tcl8 -c++   -o example_wrap.cxx example.i
g++ -c -fpic -I. -I/include/boost-0    example_wrap.cxx  -isystem /usr/include
In file included from /usr/include/c++/10.1.0/ext/string_conversions.h:41,
                 from /usr/include/c++/10.1.0/bits/basic_string.h:6535,
                 from /usr/include/c++/10.1.0/string:55,
                 from /usr/include/c++/10.1.0/stdexcept:39,
                 from example_wrap.cxx:1683:
/usr/include/c++/10.1.0/cstdlib:75:15: fatal error: stdlib.h: No such file or directory
   75 | #include_next <stdlib.h>
      |               ^~~~~~~~~~
compilation terminated.
make[3]: *** [../../Makefile:197: tcl_cpp] Error 1
make[3]: Leaving directory '/home/anatol/sources/swig/Examples/tcl/std_vector'
make[2]: *** [Makefile:14: build] Error 2
make[2]: Target 'check' not remade because of errors.
make[2]: Leaving directory '/home/anatol/sources/swig/Examples/tcl/std_vector'
make[1]: *** [Makefile:216: std_vector.actionexample] Error 2
```

I cd to the directory and rerun the same command
```
$ cd Examples/tcl/operator
$ g++ -c -fpic -I. -I/include/boost-0    example_wrap.cxx  -isystem /usr/include
In file included from /usr/include/c++/10.1.0/ext/string_conversions.h:41,
                 from /usr/include/c++/10.1.0/bits/basic_string.h:6535,
                 from /usr/include/c++/10.1.0/string:55,
                 from /usr/include/c++/10.1.0/stdexcept:39,
                 from example_wrap.cxx:1673:
/usr/include/c++/10.1.0/cstdlib:75:15: fatal error: stdlib.h: No such file or directory
   75 | #include_next <stdlib.h>
      |               ^~~~~~~~~~
compilation terminated.
```

But if I remove `-isystem /usr/include` part then compilation succeeds:
```
 g++ -c -fpic -I. -I/include/boost-0    example_wrap.cxx
```

And BTW there is no `/include/boost-0` directory at my system and I am not sure why the build system adds it to the search path. In fact `g++ -c -fpic example_wrap.cxx` is enough to compile this file.

Thanks @vadz for the pointer.

I went ahead and made the local changes to avoid `-isystem` https://github.com/anatol/swig/commit/d6b10e3867548454a9fcd9d8feb31730ec77cd05 . With given change the `stdlib.h` is gone.

Thank a lot!


Now I have
```
checking Examples/python/varargs
example_wrap.c:154:11: fatal error: Python.h: No such file or directory
  154 | # include <Python.h>
      |           ^~~~~~~~~~
```

The error is interesting. Especially as `./configure` detects my Python3 correctly and correctly figures out its include paths. But for some reason it does not use these python dirs for `make check`. I had to use `PY3` option like this `make check PY3=y` to build tests. With this option the python tests compile correctly and pass.

For me it makes more sense if `make check` is consistent with `./configure` and use the same default python version.

--- a/configure.ac
+++ b/configure.ac
@@ -115,15 +115,6 @@ dnl Some test cases require Boost
 AX_BOOST_BASE(,,,)
 AC_SUBST(BOOST_CPPFLAGS)
 
-dnl How to specify include directories that may be system directories.
-# -I should not be used on system directories (GCC)
-if test "$GCC" = yes; then
-    ISYSTEM="-isystem "
-else
-    ISYSTEM="-I"
-fi
-AC_MSG_NOTICE(ISYSTEM: $ISYSTEM)
-
 dnl Info for building shared libraries ... in order to run the examples
 
 # SO is the extension of shared libraries (including the dot!)
@@ -484,7 +475,7 @@ AC_ARG_WITH(tcl,
  [  --with-tcl=path         Set location of Tcl package],[
 	TCLPACKAGE="$withval"], [TCLPACKAGE="$alllang_default"])
 AC_ARG_WITH(tclincl,[  --with-tclincl=path     Set location of Tcl include directory],[
-	TCLINCLUDE="$ISYSTEM$withval"], [TCLINCLUDE=])
+	TCLINCLUDE="-I$withval"], [TCLINCLUDE=])
 AC_ARG_WITH(tcllib,[  --with-tcllib=path      Set location of Tcl library directory],[
 	TCLLIB="-L$withval"], [TCLLIB=])
 
@@ -526,7 +517,7 @@ else
     AC_MSG_RESULT(found $TCLCONFIG/tclConfig.sh)
     . $TCLCONFIG/tclConfig.sh
     if test -z "$TCLINCLUDE"; then
-        TCLINCLUDE=`echo $TCL_INCLUDE_SPEC | sed "s/-I/$ISYSTEM/"`
+        TCLINCLUDE=`echo $TCL_INCLUDE_SPEC`
     fi
     if test -z "$TCLLIB"; then
         TCLLIB=$TCL_LIB_SPEC
@@ -535,7 +526,7 @@ fi
 
 if test -z "$TCLINCLUDE"; then
    if test "x$TCLPACKAGE" != xyes; then
-	TCLINCLUDE="$ISYSTEM$TCLPACKAGE/include"
+	TCLINCLUDE="-I$TCLPACKAGE/include"
    fi
 fi
 
@@ -553,7 +544,7 @@ if test -z "$TCLINCLUDE"; then
 	for i in $dirs ; do
 		if test -r $i/tcl.h; then
 			AC_MSG_RESULT($i)
-			TCLINCLUDE="$ISYSTEM$i"
+			TCLINCLUDE="-I$i"
 			break
 		fi
 	done
@@ -1004,7 +995,7 @@ if test -n "$PERL"; then
 			AC_MSG_RESULT($PERL5LIB)
 		fi
     AC_MSG_CHECKING(for Perl5 ccflags)
- 		PERL5CCFLAGS=`($PERL -e 'use Config; print $Config{ccflags}, "\n"' | sed "s/-Wdeclaration-after-statement//" | sed "s/-I/$ISYSTEM/") 2>/dev/null`
+ 		PERL5CCFLAGS=`($PERL -e 'use Config; print $Config{ccflags}, "\n"' | sed "s/-Wdeclaration-after-statement//") 2>/dev/null`
  		if test -z "$PERL5CCFLAGS" ; then
  			AC_MSG_RESULT(not found)
  		else
@@ -2326,7 +2317,7 @@ if test "$LUABIN"; then
   # look for the header files & set LUAFLAGS accordingly
   # will clear LUABIN if not present
   if test -n "$LUAINCLUDE"; then
-    AC_CHECK_FILE($LUAINCLUDE/lua.h,[LUAFLAGS="$ISYSTEM$LUAINCLUDE"],[LUABIN=])
+    AC_CHECK_FILE($LUAINCLUDE/lua.h,[LUAFLAGS="-I$LUAINCLUDE"],[LUABIN=])
   else
     LUA_OK="1"
     CFLAGS_SAVED=$CFLAGS
@@ -2350,7 +2341,7 @@ if test "$LUABIN"; then
         #echo "$i"
         if test -r $i/lua.h; then
           AC_MSG_RESULT($i/lua.h)
-          LUAFLAGS="$ISYSTEM$i"
+          LUAFLAGS="-I$i"
           break
         fi
       done

